File: backend/backend/src/main/java/com/anandh/mes/service/BatchService.java
package com.anandh.mes.service;

import java.util.List;

import org.springframework.data.domain.Page;

import com.anandh.mes.dto.BatchDTO;
import com.anandh.mes.enums.BatchStatus;

public interface BatchService {

    // ==========================================================
    // CRUD
    // ==========================================================

    BatchDTO createBatch(
            BatchDTO batchDTO);

    List<BatchDTO> getAllBatches();

    BatchDTO getBatchById(
            Long id);

    BatchDTO updateBatch(
            Long id,
            BatchDTO batchDTO);

    void deleteBatch(
            Long id);

    // ==========================================================
    // SEARCH
    // ==========================================================

    BatchDTO getByBatchNumber(
            String batchNumber);

    List<BatchDTO> getByStatus(
            BatchStatus status);

    List<BatchDTO> getByProductionOrder(
            Long productionOrderId);

    List<BatchDTO> getByRecipe(
            Long recipeId);

    Page<BatchDTO> getBatchPage(
            int page,
            int size,
            String sortBy);

    // ==========================================================
    // BATCH WORKFLOW
    // ==========================================================

    BatchDTO markReady(
            Long batchId);

    BatchDTO startBatch(
            Long batchId);

    BatchDTO holdBatch(
            Long batchId);

    BatchDTO resumeBatch(
            Long batchId);

    BatchDTO completeBatch(
            Long batchId,
            Double actualQuantity);

    BatchDTO closeBatch(
            Long batchId);

    BatchDTO cancelBatch(
            Long batchId);

}

File: frontend/pharma-mes-frontend/src/services/batchService.js
import api from "./api";

export const getAllBatches = async () => {
    const response = await api.get("/api/batches");
    return response.data;
};

export const getBatchById = async (id) => {
    const response = await api.get(`/api/batches/${id}`);
    return response.data;
};

export const createBatch = async (batchData) => {
    const response = await api.post(
        "/api/batches",
        batchData
    );

    return response.data;
};

export const updateBatch = async (id, batchData) => {
    const response = await api.put(
        `/api/batches/${id}`,
        batchData
    );

    return response.data;
};

export const deleteBatch = async (id) => {
    await api.delete(`/api/batches/${id}`);
};

export const markBatchReady = async (id) => {
    const response = await api.put(
        `/api/batches/${id}/ready`
    );

    return response.data;
};

export const startBatch = async (id) => {
    const response = await api.put(
        `/api/batches/${id}/start`
    );

    return response.data;
};

export const holdBatch = async (id) => {
    const response = await api.put(
        `/api/batches/${id}/hold`
    );

    return response.data;
};

export const resumeBatch = async (id) => {
    const response = await api.put(
        `/api/batches/${id}/resume`
    );

    return response.data;
};

export const completeBatch = async (
    id,
    actualQuantity
) => {
    const response = await api.put(
        `/api/batches/${id}/complete`,
        null,
        {
            params: {
                actualQuantity
            }
        }
    );

    return response.data;
};

export const closeBatch = async (id) => {
    const response = await api.put(
        `/api/batches/${id}/close`
    );

    return response.data;
};

export const cancelBatch = async (id) => {
    const response = await api.put(
        `/api/batches/${id}/cancel`
    );

    return response.data;
};

export default {
    getAllBatches,
    getBatchById,
    createBatch,
    updateBatch,
    deleteBatch,
    markBatchReady,
    startBatch,
    holdBatch,
    resumeBatch,
    completeBatch,
    closeBatch,
    cancelBatch
};